### ISSUE TITLE
ShyLU: Build errors relating to use of Kokkos tasks

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#1835]
@trilinos/shylu @trilinos/amesos2 
```
.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(700): error: name followed by "::" must be a class or namespace name
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(702): error: qualified name is not allowed
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(702): error: expected a ";"
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(708): error: identifier "sched_type_host" is undefined
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(710): error: identifier "future_type" is undefined
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(721): error: identifier "sched_type_host" is undefined
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

...Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(772): error: namespace "Kokkos" has no member "host_spawn"
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(772): error: namespace "Kokkos" has no member "TaskSingle"
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
/scratch/prj/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(772): error: name followed by "::" must be a class or namespace name
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(774): error: namespace "Kokkos" has no member "wait"
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(499): error: name followed by "::" must be a class or namespace name
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

[ 55%] Built target Anasazi_LOBPCG_complex_lap_test
.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(503): error: identifier "sched_type_host" is undefined
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here
.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(515): error: identifier "sched_type_host" is undefined
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(563): error: namespace "Kokkos" has no member "host_spawn"
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(563): error: namespace "Kokkos" has no member "TaskSingle"
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(563): error: name followed by "::" must be a class or namespace name
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here

.../Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp(565): error: namespace "Kokkos" has no member "wait"
          detected during:
            instantiation of "int Tacho::Solver<ValueType, DeviceSpaceType>::factorize(const Tacho::Solver<ValueType, DeviceSpaceType>::value_type_array &) [with ValueType=double, DeviceSpaceType=Kokkos::DefaultHostExecutionSpace]"
.../Trilinos/packages/amesos2/src/Amesos2_Tacho_def.hpp(190): here
            instantiation of "int Amesos2::TachoSolver<Matrix, Vector>::numericFactorization_impl() [with Matrix=Epetra_CrsMatrix, Vector=Epetra_MultiVector]"
.../Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp(153): here
```

**Comment by kyungjoo-kim** [comment]
how do you get this error ? Can you put a configure script ?

**Comment by kyungjoo-kim** [comment]
Could you try this configure script ? 
```
CMAKE_PROJECT_DIR=/home/kyukim/Work/lib/trilinos/temp
CMAKE_INSTALL_DIR=/home/kyukim/Work/lib/trilinos/install/issue-1835
cmake \
    -DTrilinos_ENABLE_TESTS:BOOL=ON \
    -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF \
    -DTPL_ENABLE_MPI=ON \
    -DCMAKE_BUILD_TYPE=RELEASE \
    -DTrilinos_ENABLE_DEBUG=ON \
    -DBUILD_SHARED_LIBS=ON \
    -DTrilinos_ENABLE_Amesos2:BOOL=ON \
    -DTrilinos_ENABLE_ShyLU_NodeTacho:BOOL=ON \
    -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
    -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=OFF \
    -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_DIR} \
    ${CMAKE_PROJECT_DIR}
```

I was able to compile it successfully and pass ctests.
```
[kyukim @bread] issue-1835 > ctest
Test project /home/kyukim/Work/lib/trilinos/build/issue-1835
    Start 1: ShyLU_NodeTacho_Tacho_TestUtil_MPI_1
1/9 Test #1: ShyLU_NodeTacho_Tacho_TestUtil_MPI_1 .................   Passed    0.18 sec
    Start 2: ShyLU_NodeTacho_Tacho_TestSerial_double_MPI_1
2/9 Test #2: ShyLU_NodeTacho_Tacho_TestSerial_double_MPI_1 ........   Passed    0.10 sec
    Start 3: ShyLU_NodeTacho_Tacho_TestSerial_dcomplex_MPI_1
3/9 Test #3: ShyLU_NodeTacho_Tacho_TestSerial_dcomplex_MPI_1 ......   Passed    0.10 sec
    Start 4: Amesos2_KLU2_UnitTests_MPI_2
4/9 Test #4: Amesos2_KLU2_UnitTests_MPI_2 .........................   Passed    1.36 sec
    Start 5: Amesos2_Tacho_Solver_Test_MPI_1
5/9 Test #5: Amesos2_Tacho_Solver_Test_MPI_1 ......................   Passed    0.26 sec
    Start 6: Amesos2_SolverFactory_UnitTests_MPI_4
6/9 Test #6: Amesos2_SolverFactory_UnitTests_MPI_4 ................   Passed    1.26 sec
    Start 7: Amesos2_Tpetra_MultiVector_Adapter_UnitTests_MPI_4
7/9 Test #7: Amesos2_Tpetra_MultiVector_Adapter_UnitTests_MPI_4 ...   Passed    1.27 sec
    Start 8: Amesos2_Tpetra_CrsMatrix_Adapter_UnitTests_MPI_4
8/9 Test #8: Amesos2_Tpetra_CrsMatrix_Adapter_UnitTests_MPI_4 .....   Passed    1.30 sec
    Start 9: Amesos2_SimpleSolveExampleNonContigMap_MPI_2
9/9 Test #9: Amesos2_SimpleSolveExampleNonContigMap_MPI_2 .........   Passed    1.25 sec

100% tests passed, 0 tests failed out of 9

Label Time Summary:
Amesos2       =   6.69 sec (6 tests)
ShyLU_Node    =   0.39 sec (3 tests)

Total Test time (real) =   7.10 sec
```

**Comment by mhoemmen** [comment]
@kyungjoo-kim My build was a CUDA build, through the check-in test script, using the `--extra-builds` flag to enable all packages.  Could that be the issue?  I got around this by disabling the `ShyLU` and `ShyLU_Node` packages.

**Comment by mhoemmen** [comment]
@kyungjoo-kim Just curious: Do you actually need `Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF`?

**Comment by kyungjoo-kim** [comment]
@mhoemmen 

The solver does not supports CUDA (no graph partitioner, no blas library, etc). Even if the code is enalbed via UVM, the host execution space should be used. Maybe we need to disable the amesos2 wrapper for CUDA. Can this be a solution ? 

@MicheldeMessieres 

Do you have any idea of this ?

**Comment by mhoemmen** [comment]
@kyungjoo-kim That's fine, but it would be nice if it _built_ with CUDA :-) .  I'm OK if you just defer to the host execution space in that case.

**Comment by kyungjoo-kim** [comment]
@mhoemmen I think that this is resolved by #2580 . Can we close this ?

**Comment by mhoemmen** [comment]
@kyungjoo-kim I'm ok with that -- thanks!
